### PR TITLE
Owning fields should not be compared with string containment

### LINKED ISSUE #6276: Resource Leak Checker: false negative related to owning fields
Based on my review of #6271, I wrote the following test case, which demonstrates an unsoundness in the current implementation of the Resource Leak Checker. The same problem exists in the code in #6271, but @Calvin-L has helpfully added a "TODO: this is very wrong" comment to the code that's responsible (thanks to @Calvin-L for that - I would not have noticed the issue otherwise):

```
// Test case for https://github.com/typetools/checker-framework/issues/6276

import org.checkerframework.checker.mustcall.qual.*;
import org.checkerframework.checker.calledmethods.qual.*;
import java.net.Socket;

@InheritableMustCall("a")
public class OwningFieldStringComparison {

  // :: error: required.method.not.called
  @Owning Socket s;

  // important to the bug: the name of this field must contain
  // the name of the owning socket
  /* @NotOwning */ Socket s2;

  // Note this "destructor" closes the wrong socket
  @EnsuresCalledMethods(value="this.s2", methods="close")
  public void a() {
    try {
      this.s2.close();
    } catch(Exception e) {

    } finally {
      try {
        this.s2.close();
      } catch (Exception e) {

      }
    }
  }
}
```

This problem is that `String#contains` is used by the RLC internally to check whether the `value` field of a destructor's `@EnsuresCalledMethods` annotation contains the simple name of an owning field. Instead, something like program element equality should be checked, instead.

### PR DESCRIPTION
**Pull request by kelloggm**
fixes #6276

sorry in advance for the mildly annoying merge with #6271 @Calvin-L

### CODE REVIEW

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java [hunk 1]
```diff
@@ -467,6 +471,23 @@ private void checkOwningField(Element field) {
...
+   * Determine if the given expression <code>e</code> refers to <code>this.field</code>.
+   *
+   * @param e the expression
+   * @param field the field
+   * @return true if <code>e</code> refers to <code>this.field</code>
+   */
+  private boolean expressionEqualsField(String e, VariableElement field) {
+    try {
+      JavaExpression je = StringToJavaExpression.atFieldDecl(e, field, this.checker);
+      return je instanceof FieldAccess && ((FieldAccess) je).getField().equals(field);
```
Comment (Calvin-L) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java:484]: Do we also need to check if the receiver is `this`?

**Comment by Calvin-L** [comment]
> sorry in advance for the mildly annoying merge

It's no problem. I'm happy to see this fixed. :)

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java (hunk 1 above)
Comment (kelloggm) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java:484]: `StringToJavaExpression` should handle that for us: it uses the supplied location (`field`, the declaration of the owning field, in this case) and the given string to find the appropriate program element, and associates that with the result.

A potential point of confusion in my implementation is that `field` is used for two purpose:
* it's the element corresponding to the owning field, so the object that `StringToJavaExpression` converts the input string into is compared against it
* it is used as the location at which `StringToJavaExpression` should consider the string `e` to occur in the code: that is, the context

Comment (Calvin-L) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java:484]: Could I write something like

```java
@MustCall("close")
class SneakyDestructor {
  private final @Owning Closeable resource;

  // ...

  @EnsuresCalledMethods(value = "#1.resource", methods = "close")
  void close(SneakyDestructor other) {
    other.close();
  }
}
```

?

In that case, the field referred to by the expression really is `SneakyDestructor.resource`, but the destructor clearly doesn't close its own `@Owning` field.

Comment (kelloggm) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java:484]: I understand what you mean now, thanks. I fleshed out your test case and added it to this PR, and a bit to my surprise the tests are passing.

**Review by msridhar — APPROVED** [review]
LGTM, nice fix!

**Pull request merged**